### 개요

---

MVP로 맡은 부분인 결제 부분에서 책에 대한 동시성 제어를 진행을 맡아 진행을 하였다.

다른 블로그를 참고를 해가면서 만들었지만 레이스 컨디션이 발생하는 것을 확인을 하였다.

이 문제를 해결하면서 생긴 과정에 대해서 다루어 보겠다.

### 문제상황

---

[인공지능 맞춤 추천 (1)](https://www.notion.so/1-1e62dc3ef51480c7b407e7e02dc9c233?pvs=21)

- 소스코드
    
    ```jsx
    @Transactional
    	public void payment(/*Long userId, */Long bookId, Long buyStock, Long money, PayType payType) {
    //		UserResponse user = userService.getMyInfo(userId);
    
    		Book book = bookRepository.findById(bookId).orElseThrow(()->new NotFoundException("Book not found"));
    
    		RLock lock = redissonClient.getFairLock("book:"+bookId);
    
    		try {
    			boolean acquired = lock.tryLock(100L, 10L, TimeUnit.SECONDS);
    			if (!acquired) {
    				throw new InterruptedException();
    			}
    			
    			// 책이 재고가 0개인 경우 || 구매하려는 개수 만큼 없는 경우
    			if( book.getStock() == 0 || book.getStock() < buyStock ) {
    				throw new InvalidRequestException(ErrorMessage.ZERO_BOOK_STOCK.getMessage());
    			}
    
    			// 돈이 부족한 경우
    			if( money < buyStock * book.getPrePrice() ) {
    				throw new InvalidRequestException(ErrorMessage.SHORT_ON_MONEY.getMessage());
    			}
    
    			// buyStock 만큼 구매
    			book.updateStock(book.getStock() - buyStock);
    
    			bookRepository.save(book);
    
    			// todo : 오더 로직
    
    		} catch (InterruptedException ex) {
    			throw new InvalidRequestException(ErrorMessage.REDIS_ERROR.getMessage());
    		} catch ( InvalidRequestException ex ) {
    			throw new InvalidRequestException(ex.getMessage());
    		} catch (Exception ex) {
    			throw new InvalidRequestException(ErrorMessage.ERROR.getMessage());
    		} finally {
    			lock.unlock();
    		}
    	}
    ```
    
- 테스트코드
    
    ```jsx
    @ExtendWith(MockitoExtension.class)
    class PaymentServiceTest {
    
    	@Mock
    	private BookRepository bookRepository;
    
    	@InjectMocks
    	private PaymentService paymentService;
    
    	@Mock
    	private RedissonClient redissonClient;
    
    	@Mock
    	private RLock rlock;
    
    	@Test
    	@DisplayName("단일 호출 테스트")
    	void paymentTest() throws InterruptedException {
    		//given
    		Book book = new Book();
    		ReflectionTestUtils.setField(book, "bookId", 1L);
    		ReflectionTestUtils.setField(book, "count", 1000L);
    
    		given(bookRepository.findById(anyLong())).willReturn(Optional.of(book));
    		given(redissonClient.getFairLock("book:"+book.getBookId())).willReturn(rlock);
    		given(rlock.tryLock(10L, 1L, TimeUnit.SECONDS)).willReturn(true);
    
    		// when
    		paymentService.payment(1L);
    
    		// then
    		verify(bookRepository).findById(1L);
    		assertEquals(999, book.getCount());
    	}
    
    	@Test
    	@DisplayName("멀티 쓰래드 테스트")
    	void paymentTest_thread() throws InterruptedException {
    		//given
    		Book book = new Book();
    		ReflectionTestUtils.setField(book, "bookId", 1L);
    		ReflectionTestUtils.setField(book, "count", 1000L);
    
    		int threadCount = 1000;
    		ExecutorService executorService = Executors.newFixedThreadPool(threadCount);
    		CountDownLatch latch = new CountDownLatch(threadCount);
    
    		given(redissonClient.getFairLock("book:"+book.getBookId())).willReturn(rlock);
    		given(bookRepository.findById(anyLong())).willReturn(Optional.of(book));
    		given(rlock.tryLock(10L, 1L, TimeUnit.SECONDS)).willReturn(true);
    
    		//when
    		for (int i = 0; i < threadCount; i++) {
    			executorService.execute(() -> {
    				try {
    					paymentService.payment(1L);
    				} catch (Exception e) {
    					System.out.println(e.getMessage());
    				} finally {
    					latch.countDown();
    				}
    			});
    		}
    
    		latch.await();
    		executorService.shutdown();
    
    		assertEquals(0, book.getCount());
    	}
    }
    ```
    

코드에서는 레디스로 fairLock을 획득해오고, 책을 가져오고 해당 내용을 한 권을 빼고 진행을 한다.

마지막으로 락을 해제해주는데, 로직상 보면 문제가 없다고 생각을 하였다 이제 원인을 분석을 해보도록 하겠다.

### 원인 분석

---

- 분산락이 적용이 되어있는지 확인을 해보았다.
    1. 일반락
        
        ![스크린샷 2025-04-15 15-55-14.png](attachment:9628749f-cdce-495d-b415-2ef62cc02691:스크린샷_2025-04-15_15-55-14.png)
        
        ![스크린샷 2025-04-15 15-55-32.png](attachment:06ea140d-4580-4103-9cd8-b884fd04c948:스크린샷_2025-04-15_15-55-32.png)
        
    2. 분산락
        
        ![img (4).png](attachment:200313ba-4a18-4917-8ad9-67b3f9668b58:img_(4).png)
        
        ![다운로드 (1).png](attachment:6efbb9a8-3a6e-4ab7-8a5f-352c94390764:다운로드_(1).png)
        

일반과 비교를 했을 때 확실하게 분산락이 적용이 되어있는 점을 확인을 하였다.

- 트랜잭션 적용

문제는 이것이였는데 트랜잭션이 적용이 되어있어서 생긴 문제였다.

### 해결 방법

---

문제 해결 방법은 매우 간단하였다.

1. book 객체를 가져오는 것을 락 안쪽에 넣을 것.
2. 트랜잭션을 빼는 것이였다.

book 객체를 락 밖에서 가져오면 동시성 제어가 안되는 상태로 가져오는 것이기 때문에 문제가 생겼던 부분이였다.

두번째로 트랜잭션인데, 트랜잭션이 문제인 것을 안 것이 트랜잭션을 빼고, 돌려보니 해결이 되었었다.

트랜잭션 내부를 어떻게 돌아가지는지 확인은 못하였지만,

문제를 예상하기로는 트랜잭션과 락의 생명주기가 달라서 생긴 문제가 아니였을까 싶다.

락의 생명주기 보다 트랜잭션의 생명주기가 길어서 DB에 적용이 되기 전에 락의 생명주기가 끝나 다른 정보를 주고 처리를 하게 되지 않았을까 생각을 한다.

### 적용

---

- 수정된 소스코드
    
    ```jsx
    public void payment(Long userId, Long bookId, Long buyStock, Long money, PayType payType) {
    		UserResponse user = userService.getMyInfo(userId);
    
    		RLock lock = redissonClient.getFairLock("book:"+bookId);
    
    		try {
    			boolean acquired = lock.tryLock(100L, 10L, TimeUnit.SECONDS);
    			if (!acquired) {
    				throw new InterruptedException();
    			}
    			Book book = bookRepository.findById(bookId).orElseThrow(()->new NotFoundException("Book not found"));
    			// 책이 재고가 0개인 경우 || 구매하려는 개수 만큼 없는 경우
    			if( book.getStock() == 0 || book.getStock() < buyStock ) {
    				throw new InvalidRequestException(ErrorMessage.ZERO_BOOK_STOCK.getMessage());
    			}
    
    			// 돈이 부족한 경우
    			if( money < buyStock * book.getPrePrice() ) {
    				throw new InvalidRequestException(ErrorMessage.SHORT_ON_MONEY.getMessage());
    			}
    
    			// buyStock 만큼 구매
    			book.updateStock(book.getStock() - buyStock);
    
    			bookRepository.save(book);
    
    			// todo : 오더 로직
    
    		} catch (InterruptedException ex) {
    			throw new InvalidRequestException(ErrorMessage.REDIS_ERROR.getMessage());
    		} catch ( InvalidRequestException ex ) {
    			throw new InvalidRequestException(ex.getMessage());
    		} catch (Exception ex) {
    			throw new InvalidRequestException(ErrorMessage.ERROR.getMessage());
    		} finally {
    			lock.unlock();
    		}
    	}
    ```
    

### 문제 해결 및 결과

---

![스크린샷 2025-05-01 오후 7.01.49.png](attachment:6a92c262-c82e-4b75-bace-6d74eddc290e:스크린샷_2025-05-01_오후_7.01.49.png)

![다운로드 (2).png](attachment:1a3a0f38-b072-4bea-8127-90ea4501d58b:다운로드_(2).png)

![다운로드 (3).png](attachment:fc97b5ee-1b15-4e3d-b385-cf15071582bf:다운로드_(3).png)

트랜잭션과 book 찾는 순서를 변경하니 레이스 컨디션이 해결된 모습을 확인을 할 수 있었다.
